include/sparrow/layout.hpp

```diff
@@ -0,0 +1,251 @@
+    {
+    }
+
+    template <class T, bool is_const>
+    auto primitive_layout_iterator<T, is_const>::dereference() const -> reference
+    {
+        return m_pointer[m_index];
+    }
+
+    template <class T, bool is_const>
+    void primitive_layout_iterator<T, is_const>::increment()
+    {
+        ++m_index;
+    }
+
+    template <class T, bool is_const>
+    void primitive_layout_iterator<T, is_const>::decrement()
+    {
+        --m_index;
+    }
+
+    template <class T, bool is_const>
+    void primitive_layout_iterator<T, is_const>::advance(difference_type n)
+    {
+        m_index += n;
+    }
+
+    template <class T, bool is_const>
+    auto primitive_layout_iterator<T, is_const>::distance_to(const self_type& rhs) const -> difference_type
+    {
+        return rhs.m_index - m_index;
+    }
+
+    template <class T, bool is_const>
+    bool primitive_layout_iterator<T, is_const>::equal(const self_type& rhs) const
+    {
+        return distance_to(rhs) == 0;
+    }
+
+    template <class T, bool is_const>
+    bool primitive_layout_iterator<T, is_const>::less_than(const self_type& rhs) const
+    {
+        return distance_to(rhs) > 0;
+    }
+
+    /***********************************
+     * primitive_layout implementation *
+     * ********************************/
+
+    template <class T>
+    primitive_layout<T>::primitive_layout(array_data ad)
+        : m_data(ad)
+    {
+        if (m_data.buffers.size() == 0)
+        {
+            throw std::runtime_error("No buffers are present in array_data");
+        }
+    }
+
+    template <class T>
+    auto primitive_layout<T>::size() const -> size_type
+    {
+        return m_data.buffers[0].size();
+    }
+
+    template <class T>
+    auto primitive_layout<T>::element(size_type i) -> reference
+    {
+        assert(pos < size());
+        return i[data()];
+    }
+
+    template <class T>
+    auto primitive_layout<T>::element(size_type i) const -> const_reference
+    {
+        assert(pos < size());
+        return i[data()];
+    }
+
+    template <class T>
+    auto primitive_layout<T>::begin() -> iterator
+    {
+        return iterator(data());
+    }
+
+    template <class T>
+    auto primitive_layout<T>::end() -> iterator
+    {
+        return iterator(data()) + self_type::size();
+    }
+
+    template <class T>
+    auto primitive_layout<T>::begin() const -> const_iterator
+    {
+        return cbegin();
+    }
+
+    template <class T>
+    auto primitive_layout<T>::end() const -> const_iterator
+    {
+        return cend();
+    }
+
+    template <class T>
+    auto primitive_layout<T>::cbegin() const -> const_iterator
+    {
+        return const_iterator(data());
+    }
+
+    template <class T>
+    auto primitive_layout<T>::cend() const -> const_iterator
+    {
+        return const_iterator(data()) + self_type::size();
+    }
+
+    template <class T>
+    auto primitive_layout<T>::data() -> pointer
+    {
+        return m_data.buffers[0].data();
```

Comment:
This assumes that `buffers.size() > 0` but my understanding is that it's not always the case?
@JohanMabille You know best what's required here exactly I think.

Reply:
for primitive layout, it is expected that `m_data.buffers.size() == 1`, we should add an assertion in the constructor.

Reply:
The assertion shoudl be here too. There is a situation where the buffer becomes empty: when we pass the ownership to something else.

Reply:
For now an exception is thrown in the constructor, but should we instead just use assertions?

Reply:
Yes